### PR TITLE
Add support for dashboard json resource

### PR DESCRIPTION
**Pull request by skarimo**

### CODE REVIEW

File: datadog/resource_datadog_dashboard_json.go [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+			},
+		},
+	}
+}
+
+func resourceDatadogDashboardJSONRead(d *schema.ResourceData, meta interface{}) error {
+	providerConf := meta.(*ProviderConfiguration)
+	datadogClientV1 := providerConf.DatadogClientV1
+	authV1 := providerConf.AuthV1
+	path := "/api/v1/dashboard/" + d.Id()
```
Comment (therve) [inline datadog/resource_datadog_dashboard_json.go:60]: Can you create a constant with that path?

File: datadog/tests/resource_datadog_dashboard_json_test.go [hunk 2]
```diff
@@ -0,0 +1,1310 @@
...
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+)
+
+func TestAccDatadogDashboardJSON_BasicTimeboard(t *testing.T) {
+	ctx, accProviders := testAccProviders(context.Background(), t)
+	uniq := uniqueEntityName(ctx, t)
+	accProvider := testAccProvider(t, accProviders)
+
+	t.Parallel()
```
Comment (therve) [inline datadog/tests/resource_datadog_dashboard_json_test.go:16]: Please put that first in the test, that avoids creating providers in memory.

File: datadog/tests/resource_datadog_dashboard_json_test.go [hunk 3]
```diff
@@ -0,0 +1,1310 @@
...
+	resource.Test(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t) },
+		Providers: accProviders,
+		// Import checkDashboardDestroy() from Dashboard resource
+		CheckDestroy: checkDashboardDestroy(accProvider),
```
Comment (therve) [inline datadog/tests/resource_datadog_dashboard_json_test.go:21]: 👍

File: datadog/tests/resource_datadog_dashboard_json_test.go [hunk 4]
```diff
@@ -0,0 +1,1310 @@
...
+		},
+	})
+}
+
+func TestAccDatadogDashboardJSON_BasicScreenboard(t *testing.T) {
+	ctx, accProviders := testAccProviders(context.Background(), t)
+	uniq := uniqueEntityName(ctx, t)
+	accProvider := testAccProvider(t, accProviders)
+
+	t.Parallel()
```
Comment (therve) [inline datadog/tests/resource_datadog_dashboard_json_test.go:48]: Same here.

File: datadog/resource_datadog_dashboard_json.go [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+	return updateDashboardJSONState(d, respMap)
+}
+
+func resourceDatadogDashboardJSONCreate(d *schema.ResourceData, meta interface{}) error {
+	providerConf := meta.(*ProviderConfiguration)
+	datadogClientV1 := providerConf.DatadogClientV1
+	authV1 := providerConf.AuthV1
+	path := "/api/v1/dashboard"
+
+	test, err := utils.NormalizeJSONorYamlString(d.Get("dashboard").(string))
```
Comment (therve) [inline datadog/resource_datadog_dashboard_json.go:85]: That's an interesting approach. FWIW my thinking was more to check if calling `jsonencode(yamldecode(data)` would work. Do you think that would simplify the code?

Comment (therve) [reply datadog/resource_datadog_dashboard_json.go:85]: Also let's name the variable something else than `test`.

**Review by therve — COMMENTED** [review]
That looks great, some questions around and minor comments.

File: datadog/resource_datadog_dashboard_json.go [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+				StateFunc: func(v interface{}) string {
+					// Remove computed fields when comparing diffs
+					attrMap, _ := structure.ExpandJsonFromString(v.(string))
+					for _, f := range computedFields {
+						delete(attrMap, f)
+					}
+					res, _ := structure.FlattenJsonToString(attrMap)
+					return res
+				},
+				Description: "The JSON or YAML formatted definition of the Dashboard.",
```
Comment (zippolyte) [inline datadog/resource_datadog_dashboard_json.go:38]: Wait, is yaml supported natively ? How does that work 🤯

**Review by zippolyte — CHANGES_REQUESTED** [review]
Just a couple of comments, looks great overall

File: datadog/tests/resource_datadog_dashboard_json_test.go [hunk 7]
```diff
@@ -0,0 +1,1314 @@
...
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+)
+
+func TestAccDatadogDashboardJSON_BasicTimeboard(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (zippolyte) [reply datadog/tests/resource_datadog_dashboard_json_test.go]: Yeah we do, but I thought this was changed by Thomas when he added the linter.

[the remark this answers is not in the document]
Comment (therve) [reply datadog/tests/resource_datadog_dashboard_json_test.go]: It doesn't fail for public names, so `Test_Basic` is allowed, not `test_basic`.

**Review by zippolyte — APPROVED** [review]

**Pull request merged**